File: schema/schema.go
package schema

import (
	"fmt"
	"strings"
)

type APISchema struct {
	Type      string     `json:"type"` // "openapi", "wordpress", "graphql"
	BaseURL   string     `json:"base_url"`
	Endpoints []Endpoint `json:"endpoints"`
}

type Endpoint struct {
	Path        string            `json:"path"`
	Methods     []string          `json:"methods"`
	Description string            `json:"desciption,omitempty"`
	Args        map[string]Arg    `json:"args,omitempty"`
	Responses   map[string]string `json:"responses,omitempty"`
}

type Arg struct {
	Type     string   `json:"type,omitempty"`
	Required bool     `json:"required,omitempty"`
	In       string   `json:"in,omitempty"` // "query", "body", "path", "header"
	Enum     []string `json:"enum,omitempty"`
	Default  any      `json:"default,omitempty"`
}

func (s *APISchema) Compress() ([]string, error) {
	var compressedEndpoints []string

	for _, endpoint := range s.Endpoints {
		methods := strings.Join(endpoint.Methods, ",")
		path := endpoint.Path

		var argList []string
		for key, value := range endpoint.Args {
			enums := strings.Join(value.Enum, "|")
			requiredOrNot := ""
			if value.Required == true {
				requiredOrNot = "*"
			}

			defaultValue := ""
			if value.Default != nil {
				defaultValue = fmt.Sprintf("=%v", value.Default)
			}

			arg := fmt.Sprintf("%v%v:%v%v{%v}", requiredOrNot, key, value.Type, defaultValue, value.In)
			if len(enums) > 0 {
				arg += fmt.Sprintf("[%v]", enums)
			}

			argList = append(argList, arg)
		}

		args := strings.Join(argList, ";")

		compressedEndpoint := fmt.Sprintf("[%v] %v", methods, path)

		if len(args) > 0 {
			compressedEndpoint += fmt.Sprintf(" | %v", args)
		}

		compressedEndpoints = append(compressedEndpoints, compressedEndpoint)
	}

	result := strings.Join(compressedEndpoints, "\n")
	fmt.Println(result)
	return compressedEndpoints, nil
}

File: schema/schema.py
from dataclasses import dataclass, field
from typing import Any


@dataclass
class Arg:
    type: str = ""
    required: bool = False
    in_: str = ""  # "query", "body", "path", "header"
    enum: list[str] = field(default_factory=list)
    default: Any = None


@dataclass
class Endpoint:
    path: str = ""
    methods: list[str] = field(default_factory=list)
    description: str = ""
    args: dict[str, Arg] = field(default_factory=dict)
    responses: dict[str, str] = field(default_factory=dict)


@dataclass
class APISchema:
    type: str = ""  # "openapi", "wordpress", "graphql"
    base_url: str = ""
    endpoints: list[Endpoint] = field(default_factory=list)

    def Compress(self) -> list[str]:
        compressed_endpoints = []

        for endpoint in self.endpoints:
            methods = ",".join(endpoint.methods)
            path = endpoint.path

            arg_list = []
            for key, value in endpoint.args.items():
                enums = "|".join(value.enum)
                required_or_not = "*" if value.required else ""
                default_value = f"={value.default}" if value.default is not None else ""

                arg = f"{required_or_not}{key}:{value.type}{default_value}{{{value.in_}}}"
                if enums:
                    arg += f"[{enums}]"

                arg_list.append(arg)

            args = ";".join(arg_list)

            compressed_endpoint = f"[{methods}] {path}"
            if args:
                compressed_endpoint += f" | {args}"

            compressed_endpoints.append(compressed_endpoint)

        return compressed_endpoints


@dataclass
class Findings:
    endpoint: str = ""
    method: str = ""
    attack: str = ""
    severity: str = ""
    evidence: str = ""
    request: str = ""
    response: str = ""
